backend/dataall/modules/worksheets/services/worksheet_query_result_service.py

```diff
@@ -0,0 +1,139 @@
+        excel_buffer.seek(0)
+        return excel_buffer
+
+    @staticmethod
+    def handle_xlsx_format(output_bucket: str, file_key: str) -> bool:
+        aws_region_name = os.getenv('AWS_REGION_NAME', 'eu-west-1')
+        file_name, _ = file_key.split('.')
+        csv_data = S3_client.get_object(region=aws_region_name, bucket=output_bucket, key=f'{file_name}.csv')
+        excel_buffer = WorksheetQueryResultService.convert_csv_to_xlsx(csv_data)
+        S3_client.put_object(region=aws_region_name, bucket=output_bucket, key=file_key, body=excel_buffer)
+        return True
+
+    @staticmethod
+    def download_sql_query_result(session: 'Session', data: dict = None):
+        # # default timeout for the download link is 2 hours(in minutes)
+        default_timeout = os.getenv('QUERY_RESULT_TIMEOUT_MINUTES', 120)
+
+        environment = EnvironmentService.get_environment_by_uri(session, data.get('environmentUri'))
+        worksheet = WorksheetService.get_worksheet_by_uri(session, data.get('worksheetUri'))
+        env_group = EnvironmentService.get_environment_group(
+            session, worksheet.SamlAdminGroupName, environment.environmentUri
+        )
+        output_file_key = WorksheetQueryResultService.get_file_key(
+            env_group.environmentAthenaWorkGroup, data.get('athenaQueryId'), data.get('fileFormat')
+        )
+        sql_query_result = WorksheetRepository.find_query_result_by_format(
+            session, data.get('worksheetUri'), data.get('athenaQueryId'), data.get('fileFormat')
+        )
+        if data.get('fileFormat') == 'xlsx':
+            is_job_failed = WorksheetQueryResultService.handle_xlsx_format(
+                environment.EnvironmentDefaultBucketName, output_file_key
+            )
+            if is_job_failed:
+                raise ValueError('Error while preparing the xlsx file')
+
+        if not sql_query_result:
+            sql_query_result = WorksheetQueryResultService.create_query_result(
+                environment.EnvironmentDefaultBucketName,
+                env_group.environmentAthenaWorkGroup,
+                worksheet.worksheetUri,
+                environment.region,
+                environment.AwsAccountId,
+                data,
+            )
+        S3_client.object_exists(
+            region=environment.region, bucket=environment.EnvironmentDefaultBucketName, key=output_file_key
+        )
+        if sql_query_result.is_download_link_expired():
+            url = S3_client.get_presigned_url(
+                region=environment.region,
+                bucket=environment.EnvironmentDefaultBucketName,
+                key=output_file_key,
+                expire_minutes=default_timeout,
+            )
+            sql_query_result.downloadLink = url
+            sql_query_result.expiresIn = datetime.utcnow() + timedelta(seconds=default_timeout)
+
+        session.add(sql_query_result)
+        session.commit()
+
+        return sql_query_result
```

Comment:
Suggestion to improve download_sql_query_results (some are very nit)
1. first step, read data from db
```
        # # default timeout for the download link is 2 hours(in minutes)
        default_timeout = os.getenv('QUERY_RESULT_TIMEOUT_MINUTES', 120)

        environment = EnvironmentService.get_environment_by_uri(session, data.get('environmentUri'))
        worksheet = WorksheetService.get_worksheet_by_uri(session, data.get('worksheetUri'))
        env_group = EnvironmentService.get_environment_group(
            session, worksheet.SamlAdminGroupName, environment.environmentUri
        )
        sql_query_result = WorksheetRepository.find_query_result_by_format(
            session, data.get('worksheetUri'), data.get('athenaQueryId'), data.get('fileFormat')
        )
```
2. create sql_query_results, right after checking it and use it from now on instead of reading from data
```
        if not sql_query_result:
            sql_query_result = WorksheetQueryResultService.create_query_result(
                environment.EnvironmentDefaultBucketName,
                env_group.environmentAthenaWorkGroup,
                worksheet.worksheetUri,
                environment.region,
                environment.AwsAccountId,
                data,
            )
```
3.  build output_file_key without format to avoid creating string and then splitting it + Use enum in comparison and sql_query_results.fileFormat + avoid unnecessary middle step `handle_xlsx_format` and read directly. The if clause for `is_job_failed` won't be reached if there is an exception. 
```
        output_file_key = WorksheetQueryResultService.get_file_key(
            env_group.environmentAthenaWorkGroup, data.get('athenaQueryId')
        )

        if sql_query_results.fileFormat == WorksheetResultsFormat.XLSX.value:
            csv_data = S3_client.get_object(region=environment.region, bucket=environment.EnvironmentDefaultBucketName, key=f'{output_file_key}..{WorksheetResultsFormat.CSV.value}')
            excel_buffer = WorksheetQueryResultService.convert_csv_to_xlsx(csv_data)
            S3_client.put_object(region=environment.region, bucket=environment.EnvironmentDefaultBucketName, key=f'{output_file_key}.{WorksheetResultsFormat.XLSX.value}', body=excel_buffer)


```
4. Check if s3 file exists and if the download link is expired create a new one. Add and commit changes
```

        S3_client.object_exists(
            region=environment.region, bucket=environment.EnvironmentDefaultBucketName, key=f'{output_file_key}.{sql_query_result.fileFormat}'
        )
        if sql_query_result.is_download_link_expired():
            url = S3_client.get_presigned_url(
                region=environment.region,
                bucket=environment.EnvironmentDefaultBucketName,
                key=f'{output_file_key}.{sql_query_result.fileFormat}',
                expire_minutes=default_timeout,
            )
            sql_query_result.downloadLink = url
            sql_query_result.expiresIn = datetime.utcnow() + timedelta(seconds=default_timeout)

        session.add(sql_query_result)
        session.commit()

        return sql_query_result
```

Reply:
changes are done